<a href="https://colab.research.google.com/github/jose-gomezm/hacking-civico/blob/master/notebooks/05_Exploracion_de_datos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<p align="center">
<img src="http://codeandomexico.org/resources/img/codeandomexico.png" width="500" alt="Codeando México"><br>
<a href="http://www.codeandomexico.org/" target="_blank"><img src="https://img.shields.io/badge/website-CodeandoMexico-00D88E.svg"></a>
<a href="http://slack.codeandomexico.org/" target="_blank"><img src="https://img.shields.io/badge/slack-CodeandoMexico-EC0E4F.svg"></a>
</p>
<!-- __ -->

# Curso de Datos Abiertos y Hacking Cívico

Este curso tiene como objetivo habilitar las capacidades de la ciudadanía y los servidores públicos en el uso y generación de datos abiertos para el bien común. Puedes encontrar más información [aquí](https://github.com/CodeandoMexico/hacking-civico).

# Exploración de datos: COVID-19 en México

En este cuaderno aprenderemos algunas bases de exploración de datos. Los datos a utilizar son los datos abiertos sobre COVID-19 en México, los cuales pueden ser accedidos directamente desde el portal: https://www.gob.mx/salud/documentos/datos-abiertos-152127

Del portal de datos abiertos podemos encontrar dos enlaces, uno para descargar el conjunto de datos y otro para descargar el diccionario; así que podemos automatizar el proceso de descarga y descompresión de la información.

Para obtener los datos, sólo basta que ejecutes la siguiente línea de código, que contiene el comando de descarga de la información.

In [ ]:
# Para descargar el conjunto de datos
!wget http://datosabiertos.salud.gob.mx/gobmx/salud/datos_abiertos/datos_abiertos_covid19.zip

# Para descargar el diccionario de los datos
!wget http://epidemiologia.salud.gob.mx/gobmx/salud/datos_abiertos/diccionario_datos_covid19.zip

# Descomprimimos ambas carpetas y movemos archivos
!unzip datos_abiertos_covid19.zip
!unzip diccionario_datos_covid19.zip
!mv ./diccionario_datos_covid19/*.xlsx .
!mv *COVID19MEXICO.csv COVID19MEXICO.csv

# Eliminamos las carpetas comprimidas
!rm -rf diccionario_datos_covid19
!rm datos_abiertos_covid19.zip
!rm diccionario_datos_covid19.zip

--2020-09-02 21:46:03--  http://datosabiertos.salud.gob.mx/gobmx/salud/datos_abiertos/datos_abiertos_covid19.zip
Resolving datosabiertos.salud.gob.mx (datosabiertos.salud.gob.mx)... 187.210.186.146
Connecting to datosabiertos.salud.gob.mx (datosabiertos.salud.gob.mx)|187.210.186.146|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 28812370 (27M) [application/x-zip-compressed]
Saving to: ‘datos_abiertos_covid19.zip’

datos_abiertos_covi 100%[===================>]  27.48M  17.8MB/s    in 1.5s    

2020-09-02 21:46:05 (17.8 MB/s) - ‘datos_abiertos_covid19.zip’ saved [28812370/28812370]

--2020-09-02 21:46:05--  http://epidemiologia.salud.gob.mx/gobmx/salud/datos_abiertos/diccionario_datos_covid19.zip
Resolving epidemiologia.salud.gob.mx (epidemiologia.salud.gob.mx)... 187.191.75.207
Connecting to epidemiologia.salud.gob.mx (epidemiologia.salud.gob.mx)|187.191.75.207|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://ep

Hasta este punto, si abres la barra lateral izaquiera, deberías ser capaz de encontrar los datos:

- `COVID19MEXICO.csv`
- `Catalogos_0412.xlsx`
- `Descriptores_0419.xlsx`

👈🏼

## Carga

Ahora que tenemos los datos, procederemos a cargarlos. Para cargar los datos, primero importaremos algunos paquetes de Python, pues su funcionalidad nos será muy util.

In [ ]:
import statistics

import pandas as pd
import numpy as np

In [ ]:
# Especificamos el nombre del archivo por abrir
nombre_archivo = 'COVID19MEXICO.csv'

# Utilizamos la función de Pandas para cargar un archivo CSV
data = pd.read_csv(nombre_archivo, encoding='latin-1')

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (33) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Mostramos la cabeza (primeros elementos del archivo).

In [ ]:
data.head()

FECHA_ACTUALIZACION ID_REGISTRO  ORIGEN  ...  PAIS_NACIONALIDAD  PAIS_ORIGEN  UCI
0          2020-09-01      0b79a6       2  ...            México           99   97
1          2020-09-01      1b0c85       2  ...            México           99   97
2          2020-09-01      17ae86       2  ...            México           99    2
3          2020-09-01      14df1c       2  ...            México           99   97
4          2020-09-01      0f3272       2  ...            México           99   97

[5 rows x 35 columns]

Ahora procedemos a cargar el archivo con los descriptores.

In [ ]:
nombre_diccionario = 'Descriptores_0419.xlsx'
diccionario = pd.read_excel(nombre_diccionario)

In [ ]:
diccionario

Nº  ...                                   FORMATO O FUENTE
0    1  ...                                         AAAA-MM-DD
1    2  ...                                             TEXTO 
2    3  ...  CATÁLOGO: ORIGEN                              ...
3    4  ...  CATÁLOGO: SECTOR                              ...
4    5  ...                                CATALÓGO: ENTIDADES
5    6  ...       CATÁLOGO: SEXO                              
6    7  ...                                CATALÓGO: ENTIDADES
7    8  ...                                CATALÓGO: ENTIDADES
8    9  ...                               CATALÓGO: MUNICIPIOS
9   10  ...                           CATÁLOGO: TIPO_PACIENTE 
10  11  ...                                         AAAA-MM-DD
11  12  ...                                         AAAA-MM-DD
12  13  ...                                         AAAA-MM-DD
13  14  ...  CATÁLOGO: SI_ NO                              ...
14  15  ...  CATÁLOGO: SI_ NO                              ...
15  16  ...                                   NÚMERICA EN AÑOS
16  17  ...                            CATÁLOGO: NACIONALIDAD 
17  18  ...  CATÁLOGO: SI_ NO                              ...
18  19  ...  CATÁLOGO: SI_ NO                              ...
19  20  ...  CATÁLOGO: SI_ NO                              ...
20  21  ...  CATÁLOGO: SI_ NO                              ...
21  22  ...  CATÁLOGO: SI_ NO                              ...
22  23  ...  CATÁLOGO: SI_ NO                              ...
23  24  ...  CATÁLOGO: SI_ NO                              ...
24  25  ...  CATÁLOGO: SI_ NO                              ...
25  26  ...  CATÁLOGO: SI_ NO                              ...
26  27  ...  CATÁLOGO: SI_ NO                              ...
27  28  ...  CATÁLOGO: SI_ NO                              ...
28  29  ...  CATÁLOGO: SI_ NO                              ...
29  30  ...  CATÁLOGO: SI_ NO                              ...
30  31  ...                                CATÁLOGO: RESULTADO
31  32  ...  CATÁLOGO: SI_ NO                              ...
32  33  ...                               TEXTO, 99= SE IGNORA
33  34  ...                               TEXTO, 97= NO APLICA
34  35  ...  CATÁLOGO: SI_ NO                              ...

[35 rows x 4 columns]

Finalmente, cargamos el archivo de catálogos.

In [ ]:
nombre_cat = 'Catalogos_0412.xlsx'
sheets = ['ORIGEN', 'SECTOR', 'SEXO', 'TIPO_PACIENTE', 'SI_NO', 'NACIONALIDAD', 'RESULTADO', 'de ENTIDADES', 'MUNICIPIOS']
catalogues = {}
for sheet in sheets:
  catalogue = pd.read_excel(nombre_cat, sheet_name='Catálogo ' + sheet)
  catalogue.columns = catalogue.columns.str.replace(' ', '_')
  catalogue.columns = catalogue.columns.str.replace('.', '')
  catalogue.columns = catalogue.columns.str.lower()
  new_catalogue = {sheet: catalogue}
  catalogues.update(new_catalogue)

In [ ]:
catalogues

{'MUNICIPIOS':       clave_municipio        municipio  clave_entidad
 0                   1   AGUASCALIENTES              1
 1                   2         ASIENTOS              1
 2                   3         CALVILLO              1
 3                   4            COSÍO              1
 4                   5      JESÚS MARÍA              1
 ...               ...              ...            ...
 2496              999  NO ESPECIFICADO             32
 2497              999  NO ESPECIFICADO             36
 2498              997        NO APLICA             97
 2499              998        SE IGNORA             98
 2500              999  NO ESPECIFICADO             99
 
 [2501 rows x 3 columns], 'NACIONALIDAD':    clave      descripción
 0      1         MEXICANA
 1      2       EXTRANJERA
 2     99  NO ESPECIFICADO, 'ORIGEN':    clave      descripción
 0      1            USMER
 1      2   FUERA DE USMER
 2     99  NO ESPECIFICADO, 'RESULTADO':    clave             descripción
 0      1 

In [ ]:
catalogues['RESULTADO']

clave             descripción
0      1     Positivo SARS-CoV-2
1      2  No positivo SARS-CoV-2
2      3     Resultado pendiente

##Perfilamiento

###Descripción General

In [ ]:
# dimensión de los datos
print('Número de filas: {}'.format(data.shape[0]))
print('Número de columnas: {}'.format(data.shape[1]))

Número de filas: 1360123
Número de columnas: 35


In [ ]:
# columnas (variables)
data.columns

Index(['FECHA_ACTUALIZACION', 'ID_REGISTRO', 'ORIGEN', 'SECTOR', 'ENTIDAD_UM',
       'SEXO', 'ENTIDAD_NAC', 'ENTIDAD_RES', 'MUNICIPIO_RES', 'TIPO_PACIENTE',
       'FECHA_INGRESO', 'FECHA_SINTOMAS', 'FECHA_DEF', 'INTUBADO', 'NEUMONIA',
       'EDAD', 'NACIONALIDAD', 'EMBARAZO', 'HABLA_LENGUA_INDIG', 'DIABETES',
       'EPOC', 'ASMA', 'INMUSUPR', 'HIPERTENSION', 'OTRA_COM',
       'CARDIOVASCULAR', 'OBESIDAD', 'RENAL_CRONICA', 'TABAQUISMO',
       'OTRO_CASO', 'RESULTADO', 'MIGRANTE', 'PAIS_NACIONALIDAD',
       'PAIS_ORIGEN', 'UCI'],
      dtype='object')

In [ ]:
# limpieza del nombre de las columnas, remover espacios, carácteres especiales y pasar a minúsculas
data.columns = data.columns.str.replace(' ', '_')
data.columns = data.columns.str.replace('.', '')
data.columns = data.columns.str.lower()

In [ ]:
# calcular número de valores nulos por columna
for col in data.columns:
  print('Valores nulos en "{}": {}'.format(col,data[col].isna().sum()))

Valores nulos en "fecha_actualizacion": 0
Valores nulos en "id_registro": 0
Valores nulos en "origen": 0
Valores nulos en "sector": 0
Valores nulos en "entidad_um": 0
Valores nulos en "sexo": 0
Valores nulos en "entidad_nac": 0
Valores nulos en "entidad_res": 0
Valores nulos en "municipio_res": 0
Valores nulos en "tipo_paciente": 0
Valores nulos en "fecha_ingreso": 0
Valores nulos en "fecha_sintomas": 0
Valores nulos en "fecha_def": 0
Valores nulos en "intubado": 0
Valores nulos en "neumonia": 0
Valores nulos en "edad": 0
Valores nulos en "nacionalidad": 0
Valores nulos en "embarazo": 0
Valores nulos en "habla_lengua_indig": 0
Valores nulos en "diabetes": 0
Valores nulos en "epoc": 0
Valores nulos en "asma": 0
Valores nulos en "inmusupr": 0
Valores nulos en "hipertension": 0
Valores nulos en "otra_com": 0
Valores nulos en "cardiovascular": 0
Valores nulos en "obesidad": 0
Valores nulos en "renal_cronica": 0
Valores nulos en "tabaquismo": 0
Valores nulos en "otro_caso": 0
Valores nulos 

In [ ]:
# tipos de datos
data.dtypes

fecha_actualizacion    object
id_registro            object
origen                  int64
sector                  int64
entidad_um              int64
sexo                    int64
entidad_nac             int64
entidad_res             int64
municipio_res           int64
tipo_paciente           int64
fecha_ingreso          object
fecha_sintomas         object
fecha_def              object
intubado                int64
neumonia                int64
edad                    int64
nacionalidad            int64
embarazo                int64
habla_lengua_indig      int64
diabetes                int64
epoc                    int64
asma                    int64
inmusupr                int64
hipertension            int64
otra_com                int64
cardiovascular          int64
obesidad                int64
renal_cronica           int64
tabaquismo              int64
otro_caso               int64
resultado               int64
migrante                int64
pais_nacionalidad      object
pais_orige

###Limpieza

In [ ]:
# convertimos a fecha algunas columnas
def cast_datetime_cols(df):
    timestamp_cols = [col for col in df.columns if "fecha" in col]
    df[timestamp_cols] = df[timestamp_cols].apply(lambda date_col: pd.to_datetime(date_col, errors="coerce"), axis=0)
    return df

data = cast_datetime_cols(data)

In [ ]:
data.head()

fecha_actualizacion id_registro  origen  ...  pais_nacionalidad  pais_origen  uci
0          2020-09-01      0b79a6       2  ...            México           99   97
1          2020-09-01      1b0c85       2  ...            México           99   97
2          2020-09-01      17ae86       2  ...            México           99    2
3          2020-09-01      14df1c       2  ...            México           99   97
4          2020-09-01      0f3272       2  ...            México           99   97

[5 rows x 35 columns]

In [ ]:
data.dtypes

fecha_actualizacion    datetime64[ns]
id_registro                    object
origen                          int64
sector                          int64
entidad_um                      int64
sexo                            int64
entidad_nac                     int64
entidad_res                     int64
municipio_res                   int64
tipo_paciente                   int64
fecha_ingreso          datetime64[ns]
fecha_sintomas         datetime64[ns]
fecha_def              datetime64[ns]
intubado                        int64
neumonia                        int64
edad                            int64
nacionalidad                    int64
embarazo                        int64
habla_lengua_indig              int64
diabetes                        int64
epoc                            int64
asma                            int64
inmusupr                        int64
hipertension                    int64
otra_com                        int64
cardiovascular                  int64
obesidad    

In [ ]:
# unimos los datos con los catalogos para tener mejor entendimiento de ellos
data = (
    # unimos con el catálogo MUNICIPIOS y ENTIDADES
    data.merge(catalogues['de ENTIDADES'], how='left', left_on='entidad_res', right_on='clave_entidad').
    drop(columns=['entidad_res', 'abreviatura']).
    rename(columns={'entidad_federativa':'entidad_res'}).
    merge(catalogues['MUNICIPIOS'], how='left', left_on=['clave_entidad','municipio_res'], right_on=['clave_entidad','clave_municipio']).
    drop(columns=['municipio_res', 'clave_entidad', 'clave_municipio']).
    rename(columns={'municipio':'municipio_res'})
)

In [ ]:
# unimos con el catálogo de ENTIDADES
entidades_cols = [col for col in data.columns if "entidad" in col and col != 'entidad_res']
for col in entidades_cols:
    data = (
        data.merge(catalogues['de ENTIDADES'], how='left', left_on=col, right_on='clave_entidad').
        drop(columns=[col, 'abreviatura', 'clave_entidad']).
        rename(columns={'entidad_federativa':col})
    )

In [ ]:
# unimos con el catálogo SI_NO
sino_cols = ['intubado', 'neumonia', 'embarazo', 'habla_lengua_indig', 'diabetes', 'epoc', 'asma', 'inmusupr',
             'hipertension', 'otra_com', 'cardiovascular', 'obesidad', 'renal_cronica', 'tabaquismo', 
             'otro_caso', 'migrante', 'uci']
for col in sino_cols:
    data = (
        data.merge(catalogues['SI_NO'], how='left', left_on=col, right_on='clave').
        drop(columns=[col, 'clave']).
        rename(columns={'descripción':col})
    )

In [ ]:
# unimos otras columnnas que comparten lógica
other_cols = ['origen', 'sector', 'sexo', 'tipo_paciente', 'nacionalidad', 'resultado']
for col, cat in zip(other_cols, [x.upper() for x in other_cols]):
    data = (
        data.merge(catalogues[cat], how='left', left_on=col, right_on='clave').
        drop(columns=[col, 'clave']).
        rename(columns={'descripción':col})
    )
data.head()

fecha_actualizacion id_registro  ... nacionalidad            resultado
0          2020-09-01      0b79a6  ...     MEXICANA  Positivo SARS-CoV-2
1          2020-09-01      1b0c85  ...     MEXICANA  Positivo SARS-CoV-2
2          2020-09-01      17ae86  ...     MEXICANA  Positivo SARS-CoV-2
3          2020-09-01      14df1c  ...     MEXICANA  Positivo SARS-CoV-2
4          2020-09-01      0f3272  ...     MEXICANA  Positivo SARS-CoV-2

[5 rows x 35 columns]

In [ ]:
# creamos algunas variables que podrían ser útiles después
data['defuncion'] = ~data.fecha_def.isna()
data['mes_def'] = data.fecha_def.dt.month
data['año_def'] = data.fecha_def.dt.year
data['mes_ingreso'] = data.fecha_ingreso.dt.month
data['año_ingreso'] = data.fecha_ingreso.dt.year

In [ ]:
data.head()

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-01      0b79a6  ...           7        2020
1          2020-09-01      1b0c85  ...           4        2020
2          2020-09-01      17ae86  ...           4        2020
3          2020-09-01      14df1c  ...           4        2020
4          2020-09-01      0f3272  ...           4        2020

[5 rows x 40 columns]

###Filtros

**Distintas formas de filtrar lo mismo**

In [ ]:
# usando query
data.query('nacionalidad == "MEXICANA"').head()

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-01      0b79a6  ...           7        2020
1          2020-09-01      1b0c85  ...           4        2020
2          2020-09-01      17ae86  ...           4        2020
3          2020-09-01      14df1c  ...           4        2020
4          2020-09-01      0f3272  ...           4        2020

[5 rows x 40 columns]

In [ ]:
# haciendo referencia a la columna como índice
data[data['nacionalidad'] == 'MEXICANA'].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-01      0b79a6  ...           7        2020
1          2020-09-01      1b0c85  ...           4        2020

[2 rows x 40 columns]

In [ ]:
# seleccionando directamente la columna que queremos filtrar
data[data.nacionalidad == 'MEXICANA'].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
0          2020-09-01      0b79a6  ...           7        2020
1          2020-09-01      1b0c85  ...           4        2020

[2 rows x 40 columns]

**Filtrado múltiple**

In [ ]:
# usando query podemos poner en el string del query todas las condiciones que queramos
data.query('nacionalidad == "MEXICANA" & entidad_res == "CIUDAD DE MÉXICO"').head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
12          2020-09-01      1239f8  ...           4        2020
13          2020-09-01      0c7084  ...           5        2020

[2 rows x 40 columns]

In [ ]:
# usando el filtrado tradicional debemos encerrar en paréntesis cada condición que queremos que cumpla nuestro filtro
data[(data['nacionalidad'] == "MEXICANA") & (data.entidad_res == "CIUDAD DE MÉXICO")].head(2)

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
12          2020-09-01      1239f8  ...           4        2020
13          2020-09-01      0c7084  ...           5        2020

[2 rows x 40 columns]

###Resúmenes

In [ ]:
# una columna, una métrica
data.groupby('sexo').edad.mean()

sexo
HOMBRE    42.850075
MUJER     41.761757
Name: edad, dtype: float64

In [ ]:
# una columna, varias métricas
data.groupby('sexo').agg({'edad': ['min','mean','max']})

edad                
        min       mean  max
sexo                       
HOMBRE    0  42.850075  120
MUJER     0  41.761757  120

In [ ]:
# varias columnas, una métrica
data.groupby('sexo').agg({'edad': 'mean', 'mes_ingreso': 'mean', 'mes_def': 'mean'})

edad  mes_ingreso   mes_def
sexo                                    
HOMBRE  42.850075     6.494979  6.340483
MUJER   41.761757     6.555789  6.398543

In [ ]:
# varias columnas, varias métricas
data.groupby('sexo').agg({'edad': ['min','mean','max'], 'mes_ingreso': ['min','mean','max'], 'mes_def': ['min','mean','max']})

edad                 mes_ingreso               mes_def               
        min       mean  max         min      mean max     min      mean  max
sexo                                                                        
HOMBRE    0  42.850075  120           1  6.494979   9     1.0  6.340483  9.0
MUJER     0  41.761757  120           1  6.555789   9     1.0  6.398543  9.0

###Descripción por Tipo

In [ ]:
# Filtrar por tipos generales (categóricas y numéricas)
numerical = data.select_dtypes(include='number')
categorical = data.select_dtypes(include=['object', 'bool', 'category', 'datetime64[ns]'])

In [ ]:
numerical.head()

edad  mes_def  año_def  mes_ingreso  año_ingreso
0    30      NaN      NaN            7         2020
1    64      NaN      NaN            4         2020
2    50      4.0   2020.0            4         2020
3    27      NaN      NaN            4         2020
4    53      NaN      NaN            4         2020

In [ ]:
categorical.head()

fecha_actualizacion id_registro  ...            resultado defuncion
0          2020-09-01      0b79a6  ...  Positivo SARS-CoV-2     False
1          2020-09-01      1b0c85  ...  Positivo SARS-CoV-2     False
2          2020-09-01      17ae86  ...  Positivo SARS-CoV-2      True
3          2020-09-01      14df1c  ...  Positivo SARS-CoV-2     False
4          2020-09-01      0f3272  ...  Positivo SARS-CoV-2     False

[5 rows x 35 columns]

In [ ]:
# para conocer todos los tipo de datos
?data.select_dtypes

####Categóricas

In [ ]:
# Cardinalidad - Valores únicos por columna
for col in categorical.columns:
  print('Valores únicos en "{}": {}'.format(col,categorical[col].nunique()))

Valores únicos en "fecha_actualizacion": 1
Valores únicos en "id_registro": 1360123
Valores únicos en "fecha_ingreso": 245
Valores únicos en "fecha_sintomas": 245
Valores únicos en "fecha_def": 190
Valores únicos en "pais_nacionalidad": 119
Valores únicos en "pais_origen": 78
Valores únicos en "entidad_res": 32
Valores únicos en "municipio_res": 2153
Valores únicos en "entidad_um": 32
Valores únicos en "entidad_nac": 33
Valores únicos en "intubado": 4
Valores únicos en "neumonia": 3
Valores únicos en "embarazo": 4
Valores únicos en "habla_lengua_indig": 3
Valores únicos en "diabetes": 3
Valores únicos en "epoc": 3
Valores únicos en "asma": 3
Valores únicos en "inmusupr": 3
Valores únicos en "hipertension": 3
Valores únicos en "otra_com": 3
Valores únicos en "cardiovascular": 3
Valores únicos en "obesidad": 3
Valores únicos en "renal_cronica": 3
Valores únicos en "tabaquismo": 3
Valores únicos en "otro_caso": 3
Valores únicos en "migrante": 3
Valores únicos en "uci": 4
Valores únicos en

In [ ]:
# Unicidad - proporción de valores únicos por columna
total = data.shape[0]
for col in categorical.columns:
  print('Porcentaje de valores únicos en "{}": {}%'.format(col,round(categorical[col].nunique()/total*100,2)))

Porcentaje de valores únicos en "fecha_actualizacion": 0.0%
Porcentaje de valores únicos en "id_registro": 100.0%
Porcentaje de valores únicos en "fecha_ingreso": 0.02%
Porcentaje de valores únicos en "fecha_sintomas": 0.02%
Porcentaje de valores únicos en "fecha_def": 0.01%
Porcentaje de valores únicos en "pais_nacionalidad": 0.01%
Porcentaje de valores únicos en "pais_origen": 0.01%
Porcentaje de valores únicos en "entidad_res": 0.0%
Porcentaje de valores únicos en "municipio_res": 0.16%
Porcentaje de valores únicos en "entidad_um": 0.0%
Porcentaje de valores únicos en "entidad_nac": 0.0%
Porcentaje de valores únicos en "intubado": 0.0%
Porcentaje de valores únicos en "neumonia": 0.0%
Porcentaje de valores únicos en "embarazo": 0.0%
Porcentaje de valores únicos en "habla_lengua_indig": 0.0%
Porcentaje de valores únicos en "diabetes": 0.0%
Porcentaje de valores únicos en "epoc": 0.0%
Porcentaje de valores únicos en "asma": 0.0%
Porcentaje de valores únicos en "inmusupr": 0.0%
Porcenta

In [ ]:
# Moda - valor que más se repite en cada columna excepto los indicadores únicos
for col in categorical.drop(columns='id_registro').columns:
  print('Valor más popular en "{}": {}'.format(col, statistics.mode(categorical[col])))

Valor más popular en "fecha_actualizacion": 2020-09-01 00:00:00
Valor más popular en "fecha_ingreso": 2020-07-20 00:00:00
Valor más popular en "fecha_sintomas": 2020-07-20 00:00:00
Valor más popular en "fecha_def": NaT
Valor más popular en "pais_nacionalidad": México
Valor más popular en "pais_origen": 99
Valor más popular en "entidad_res": CIUDAD DE MÉXICO
Valor más popular en "municipio_res": IZTAPALAPA
Valor más popular en "entidad_um": CIUDAD DE MÉXICO
Valor más popular en "entidad_nac": CIUDAD DE MÉXICO
Valor más popular en "intubado": NO APLICA
Valor más popular en "neumonia": NO 
Valor más popular en "embarazo": NO 
Valor más popular en "habla_lengua_indig": NO 
Valor más popular en "diabetes": NO 
Valor más popular en "epoc": NO 
Valor más popular en "asma": NO 
Valor más popular en "inmusupr": NO 
Valor más popular en "hipertension": NO 
Valor más popular en "otra_com": NO 
Valor más popular en "cardiovascular": NO 
Valor más popular en "obesidad": NO 
Valor más popular en "r

####Numéricas

In [ ]:
numerical.describe()

edad       mes_def  año_def   mes_ingreso  año_ingreso
count  1.360123e+06  83202.000000  83202.0  1.360123e+06    1360123.0
mean   4.229901e+01      6.361554   2020.0  6.525770e+00       2020.0
std    1.677507e+01      1.200983      0.0  1.253510e+00          0.0
min    0.000000e+00      1.000000   2020.0  1.000000e+00       2020.0
25%    3.000000e+01      5.000000   2020.0  6.000000e+00       2020.0
50%    4.100000e+01      6.000000   2020.0  7.000000e+00       2020.0
75%    5.300000e+01      7.000000   2020.0  8.000000e+00       2020.0
max    1.200000e+02      9.000000   2020.0  9.000000e+00       2020.0

##Análisis Exploratorio

En esta sección responderemos algunas preguntas sobre el set de datos usando las herramientas que hemos visto hasta ahora

In [ ]:
# ¿Cuántas pruebas por origen se realizan?
data.groupby('origen').id_registro.count()

origen
FUERA DE USMER    904981
USMER             455142
Name: id_registro, dtype: int64

In [ ]:
data.origen.value_counts()

FUERA DE USMER    904981
USMER             455142
Name: origen, dtype: int64

In [ ]:
# ¿Cuántas pruebas por sector se realizan?
data.sector.value_counts()

SSA                841546
IMSS               351832
ISSSTE              51833
PRIVADA             51036
ESTATAL             25356
PEMEX               12028
IMSS-BIENESTAR       9567
SEMAR                7222
SEDENA               6735
MUNICIPAL            1114
UNIVERSITARIO         878
DIF                   769
CRUZ ROJA             199
NO ESPECIFICADO         8
Name: sector, dtype: int64

In [ ]:
# ¿Cuántos pacientes por sexo tenemos?
data.sexo.value_counts()

MUJER     688689
HOMBRE    671434
Name: sexo, dtype: int64

In [ ]:
# ¿Cuál es la proporción de los pacientes por sexo?
data.sexo.value_counts(normalize=True)

MUJER     0.506343
HOMBRE    0.493657
Name: sexo, dtype: float64

In [ ]:
# ¿Cuál es la proporción de resultados de las pruebas?
data.resultado.value_counts(normalize=True)

No positivo SARS-CoV-2    0.497718
Positivo SARS-CoV-2       0.445574
Resultado pendiente       0.056707
Name: resultado, dtype: float64

In [ ]:
# ¿Cuántas pruebas positivas tenemos en el set?
data[data.resultado == 'Positivo SARS-CoV-2'].id_registro.count()

606036

In [ ]:
# ¿Cuántos casos positivos tenemos por mes?
data[data.resultado == 'Positivo SARS-CoV-2'].mes_ingreso.value_counts().sort_index()

1         2
2         8
3      2530
4     26553
5     86407
6    152701
7    198483
8    139345
9         7
Name: mes_ingreso, dtype: int64

In [ ]:
# ¿Cuál es el resultado de las pruebas por sexo?
data.groupby(['sexo','resultado'])['id_registro'].count()

sexo    resultado             
HOMBRE  No positivo SARS-CoV-2    315398
        Positivo SARS-CoV-2       317859
        Resultado pendiente        38177
MUJER   No positivo SARS-CoV-2    361560
        Positivo SARS-CoV-2       288177
        Resultado pendiente        38952
Name: id_registro, dtype: int64

In [ ]:
data.groupby('sexo').resultado.value_counts()

sexo    resultado             
HOMBRE  Positivo SARS-CoV-2       317859
        No positivo SARS-CoV-2    315398
        Resultado pendiente        38177
MUJER   No positivo SARS-CoV-2    361560
        Positivo SARS-CoV-2       288177
        Resultado pendiente        38952
Name: resultado, dtype: int64

In [ ]:
# ¿Qué proporción de resultados de pruebas tenemos por sexo?
data.groupby('sexo').resultado.value_counts(normalize=True)

sexo    resultado             
HOMBRE  Positivo SARS-CoV-2       0.473403
        No positivo SARS-CoV-2    0.469738
        Resultado pendiente       0.056859
MUJER   No positivo SARS-CoV-2    0.524997
        Positivo SARS-CoV-2       0.418443
        Resultado pendiente       0.056560
Name: resultado, dtype: float64

In [ ]:
# ¿Cuál es la edad de los pacientes por sexo?
data.groupby('sexo').edad.describe()

count       mean        std  min   25%   50%   75%    max
sexo                                                                
HOMBRE  671434.0  42.850075  17.082438  0.0  30.0  41.0  54.0  120.0
MUJER   688689.0  41.761757  16.452140  0.0  30.0  40.0  52.0  120.0

In [ ]:
def q25(x):
  return x.quantile(0.25)

def q75(x):
  return x.quantile(0.75)

data.groupby('sexo').agg({'edad': ['min', q25, 'median', q75, 'max']})

edad                    
        min q25 median q75  max
sexo                           
HOMBRE    0  30     41  54  120
MUJER     0  30     40  52  120

In [ ]:
# ¿Cuál es la edad de los pacientes dependiendo de si son intubados o no?
data.groupby('intubado').edad.describe()

count       mean        std  min   25%   50%   75%    max
intubado                                                                      
NO                216254.0  53.623318  19.415348  0.0  42.0  55.0  67.0  120.0
NO APLICA        1115393.0  39.760982  15.004159  0.0  29.0  38.0  49.0  120.0
NO ESPECIFICADO      162.0  52.938272  17.939059  0.0  42.0  54.0  65.0   92.0
SI                 28314.0  55.728827  18.819916  0.0  47.0  58.0  68.0  108.0

In [ ]:
data.groupby('intubado').agg({'edad': ['min', q25, 'median', q75, 'max']})

edad                    
                 min q25 median q75  max
intubado                                
NO                 0  42     55  67  120
NO APLICA          0  29     38  49  120
NO ESPECIFICADO    0  42     54  65   92
SI                 0  47     58  68  108

In [ ]:
# ¿Cuál es la edad máxima y cuál es la mínima de los pacientes fallecidos?
edad_def_max = data[data.defuncion].edad.max()
data[data.defuncion & (data.edad == edad_def_max)]

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
553359          2020-09-01      1b7b20  ...           5        2020

[1 rows x 40 columns]

In [ ]:
edad_def_min = data[data.defuncion].edad.min()
data[data.defuncion & (data.edad == edad_def_min)]

fecha_actualizacion id_registro  ... mes_ingreso año_ingreso
13916            2020-09-01      02f8e2  ...           5        2020
18866            2020-09-01      083c34  ...           5        2020
40469            2020-09-01      1cc0ea  ...           4        2020
41771            2020-09-01      1b82ee  ...           5        2020
42295            2020-09-01      014b5d  ...           5        2020
...                     ...         ...  ...         ...         ...
1287531          2020-09-01      093d8f  ...           7        2020
1289734          2020-09-01      128a4d  ...           7        2020
1292236          2020-09-01      1d807b  ...           8        2020
1298812          2020-09-01      0860f4  ...           6        2020
1332531          2020-09-01      00f39c  ...           7        2020

[231 rows x 40 columns]

In [ ]:
# ¿Cuál es la edad de los pacientes que fallecen?
data.groupby('defuncion').edad.describe()

count       mean        std  min   25%   50%   75%    max
defuncion                                                               
False      1276921.0  41.028730  16.071713  0.0  30.0  40.0  51.0  120.0
True         83202.0  61.794344  15.203356  0.0  53.0  63.0  72.0  109.0

In [ ]:
# ¿Cuántas defunciones tenemos por mes?
data[data.defuncion].mes_def.value_counts().sort_index()

1.0        6
2.0       12
3.0      232
4.0     5262
5.0    16007
6.0    21298
7.0    23386
8.0    16996
9.0        3
Name: mes_def, dtype: int64

In [ ]:
# ¿Cuál es la proporción de diábeticos para las defunciones?
data.groupby('diabetes').defuncion.value_counts(normalize=True)

diabetes   defuncion
NO         False        0.956977
           True         0.043023
SE IGNORA  False        0.873052
           True         0.126948
SI         False        0.811706
           True         0.188294
Name: defuncion, dtype: float64

In [ ]:
# ¿Cuál es la proporción de asmáticos para las defunciones?
data.groupby('asma').defuncion.value_counts(normalize=True)

asma       defuncion
NO         False        0.938479
           True         0.061521
SE IGNORA  False        0.864047
           True         0.135953
SI         False        0.956997
           True         0.043003
Name: defuncion, dtype: float64

# Tarea 3

**¿Tienen los pacientes con hipertensión un riesgo más alto de defunción?**

In [ ]:
data.groupby('hipertension').defuncion.value_counts(normalize=True)

hipertension  defuncion
NO            False        0.959563
              True         0.040437
SE IGNORA     False        0.874212
              True         0.125788
SI            False        0.832883
              True         0.167117
Name: defuncion, dtype: float64

De acuerdo con los resultados, la proporción de personas con hipertensión que fallecieron a causa de COVID es de 0.1671, comparada con el 0.0404 de aquellos que no presentaban esta condición. Por ello, se puede concluir que los pacientes con hipertensión sí tienen un riesgo mayor.


**¿Cuántos casos confirmados se tienen por Estado?**

In [ ]:
positivo_covid = data.query('resultado == "Positivo SARS-CoV-2"')
positivo_covid.groupby("resultado").entidad_res.value_counts().reset_index(name='confirmados').drop(columns="resultado")

entidad_res  confirmados
0                  CIUDAD DE MÉXICO        99564
1                            MÉXICO        68516
2                        GUANAJUATO        31998
3                        NUEVO LEÓN        29524
4   VERACRUZ DE IGNACIO DE LA LLAVE        28581
5                           TABASCO        28471
6                            PUEBLA        27070
7                        TAMAULIPAS        24793
8              COAHUILA DE ZARAGOZA        21919
9                            SONORA        21659
10                          JALISCO        20595
11                  SAN LUIS POTOSÍ        18284
12                  BAJA CALIFORNIA        16987
13                          SINALOA        16244
14              MICHOACÁN DE OCAMPO        15253
15                          YUCATÁN        14923
16                         GUERRERO        14893
17                           OAXACA        13807
18                          HIDALGO        10454
19                     QUINTANA ROO        10364
20                        CHIHUAHUA         7859
21              BAJA CALIFORNIA SUR         7777
22                          DURANGO         6586
23                        QUERÉTARO         6480
24                         TLAXCALA         6479
25                          CHIAPAS         6270
26                   AGUASCALIENTES         5731
27                         CAMPECHE         5592
28                        ZACATECAS         5379
29                          MORELOS         5234
30                          NAYARIT         5037
31                           COLIMA         3713

**¿Cuántas defunciones se tienen por Estado?**

In [ ]:
data.query('defuncion == True').groupby("defuncion").entidad_res.value_counts().reset_index(name='defunciones').drop(columns="defuncion")

entidad_res  defunciones
0                            MÉXICO        12780
1                  CIUDAD DE MÉXICO        11394
2   VERACRUZ DE IGNACIO DE LA LLAVE         4553
3                            PUEBLA         4342
4                   BAJA CALIFORNIA         3896
5                           SINALOA         3503
6                           JALISCO         3498
7                            SONORA         3113
8                        NUEVO LEÓN         3088
9                           TABASCO         2802
10                       GUANAJUATO         2672
11                       TAMAULIPAS         2625
12                         GUERRERO         2184
13             COAHUILA DE ZARAGOZA         1904
14                          HIDALGO         1894
15                        CHIHUAHUA         1853
16                     QUINTANA ROO         1788
17              MICHOACÁN DE OCAMPO         1679
18                          YUCATÁN         1676
19                           OAXACA         1530
20                  SAN LUIS POTOSÍ         1502
21                         TLAXCALA         1225
22                          CHIAPAS         1207
23                          MORELOS         1189
24                         CAMPECHE          915
25                        QUERÉTARO          877
26                          NAYARIT          747
27                        ZACATECAS          654
28                          DURANGO          624
29                           COLIMA          536
30                   AGUASCALIENTES          483
31              BAJA CALIFORNIA SUR          469

**¿Cuántos fallecimientos han ocurrido en el Estado con mayor número de casos confirmados?**

In [ ]:
query = 'defuncion == True and resultado == "Positivo SARS-CoV-2"'
data.query(query).groupby("defuncion").entidad_res.value_counts().reset_index(name='defunciones').drop(columns="defuncion").iloc[0]

entidad_res    MÉXICO
defunciones     10135
Name: 0, dtype: object

| **Estado** | **Defunciones** |
|:----------:|:---------------:|
|   MEXICO   |      10135      |

**¿Cuántos fallecimientos han ocurrido en los pacientes Ambulatorios?**

In [ ]:
query = 'defuncion == True and tipo_paciente == "AMBULATORIO"'
data.query(query).defuncion.value_counts().iloc[0]

8809